src/EditorFeatures/Core/CommandHandlers/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,185 @@
+            => GetCommandState(args.SubjectBuffer, args.TextView, nextHandler);
+
+        public void ExecuteCommand(GoToNextMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: true);
+
+        public CommandState GetCommandState(GoToPreviousMethodCommandArgs args, Func<CommandState> nextHandler)
+            => GetCommandState(args.SubjectBuffer, args.TextView, nextHandler);
+
+        public void ExecuteCommand(GoToPreviousMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: false);
+
+        private static CommandState GetCommandState(ITextBuffer subjectBuffer, ITextView textView, Func<CommandState> nextHandler)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            return IsAvailable(document, caretPoint) ? CommandState.Available : nextHandler();
+        }
+
+        private static bool IsAvailable(Document document, SnapshotPoint? caretPoint)
+        {
+            var documentSupportsSuggestionService = document.Project.Solution.Workspace.Services.GetService<IDocumentSupportsSuggestionService>();
+            if (document == null || !document.SupportsSyntaxTree || !documentSupportsSuggestionService.SupportsNavigationToAnyPosition(document))
+            {
+                return false;
+            }
+
+            if (!caretPoint.HasValue)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private void ExecuteCommand(Action nextHandler, ITextBuffer subjectBuffer, ITextView textView, bool next)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            if (!IsAvailable(document, caretPoint))
+            {
+                nextHandler();
+                return;
+            }
+
+            int? targetPosition = null;
+            var waitResult = _waitIndicator.Wait(EditorFeaturesResources.Navigating, allowCancel: true, action: waitContext =>
+            {
+                targetPosition = GetTargetPosition(document, caretPoint.Value.Position, next, waitContext.CancellationToken);
+            });
+
+            if (waitResult == WaitIndicatorResult.Canceled || targetPosition == null)
+            {
+                return;
+            }
+
+            textView.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(subjectBuffer.CurrentSnapshot, targetPosition.Value), _outliningManagerService);
+        }
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal static int? GetTargetPosition(Document document, int caretPosition, bool next, CancellationToken cancellationToken)
```

Comment:
@CyrusNajmabadi: some areas somewhat like to ensure the UI thread _isn't_ pumping random other operations to avoid potential races. This is at least consistent with our pattern everywhere else.